{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# A fastq file contains the short DNA sequences that come off a sequencer. \n",
    "# Each sequence is given in a set of four lines where the second line is the sequence \n",
    "# Aand the other three lines contain information about that sequence.\n",
    "# A python script is written below to solve the given four problems regarding pattern counting\n",
    "\n",
    "# 1. Count the number of each base across all sequences.\n",
    "!curl -O http://rachelss.github.io/Bioinformatics/sample.fastq #download the sample.fastq data\n",
    "filename = 'sample.fastq' # Specify the file using a string variable\n",
    "f = open(filename,'r') # Open the file\n",
    "fastq = f.readlines() # Read it into memory, line by line\n",
    "def count_base1(file_name): # define a function with name 'count_base1', file_name refers to the data\n",
    "    count_bases = {'A':0, 'C':0, 'G':0, 'T':0} # initiate a dictionary = {key:value}\n",
    "    for line_num,line in enumerate(file_name): # include the line number using \"enumerate\" so we can find our sequences\n",
    "        if line_num % 4 == 1: # if line number / 4 leaves a remainder of 1, we can get our sequence data and we can look for patterns\n",
    "            for base in line: # access each base\n",
    "                if base == 'A': # Let's check if the base is an 'A' using an if statement\n",
    "                    count_bases['A'] += 1 # count all the A's\n",
    "                elif base == 'C':\n",
    "                      count_bases['C'] += 1 # count all the C's\n",
    "                elif base == 'G':\n",
    "                      count_bases['G'] += 1 # count all the G's\n",
    "                elif base == 'T':\n",
    "                      count_bases['T'] += 1 # count all the T's\n",
    "    print(count_bases) # print the number in the count_bases variable\n",
    "count_base1(fastq) # count_base1 function is run with fastq data\n",
    "\n",
    "\n",
    "# 2. Count the number of each pair of bases (described as kmers of size 2).\n",
    "def count_kmers2(file_name): # define a function with name 'count_kmers2', file_name refers to the data\n",
    "    count_kmers = {'AA':0, 'CC':0, 'GG':0, 'TT':0, 'AT':0, 'GT':0, 'AC':0, 'AG':0, 'CA':0, 'CG':0, 'CT':0, 'GA':0, 'GC':0, 'TA':0, 'TC':0, 'TG':0} # initiate a dictionary = {key:value}\n",
    "#count all kmers of length two in this file\n",
    "    for line_num,line in enumerate(file_name): # include the line number using \"enumerate\" so we can find our sequences\n",
    "        if line_num % 4 == 1: # if line number / 4 leaves a remainder of 1, we can get our sequence data and we can look for patterns\n",
    "            for base_num,base in enumerate (line[0:-2]):     \n",
    "                kmer = line[base_num:base_num+2]\n",
    "                kmer = ''.join(kmer) # The method join() returns a string in which the string elements of sequence have been joined by str separator.\n",
    "                count_kmers[kmer] += 1  # add the current key to the next key    \n",
    "    print(count_kmers)\n",
    "count_kmers2(fastq) # count_kmers2 function is run with fastq data\n",
    "\n",
    "\n",
    "# 3. Count the number of each pair of bases, without  assuming you know in advance the possible pairs\n",
    "def count_kmers2unknown(file_name): # define a function with name 'count_kmers2unknown', file_name refers to the data\n",
    "    count_kmers = {}\n",
    "#check if key is in dictionary or not\n",
    "    for line_num,line in enumerate(file_name): # include the line number using \"enumerate\" so we can find our sequences\n",
    "        if line_num % 4 == 1: # if line number / 4 leaves a remainder of 1, we can get our sequence data and we can look for patterns    \n",
    "            for base_num,base in enumerate (line[0:-2]):\n",
    "                kmer = line[base_num:base_num+2]\n",
    "                kmer = ''.join(kmer)  # The method join() returns a string in which the string elements of sequence have been joined by str separator.\n",
    "                if kmer in count_kmers: # check if key is in dictionary\n",
    "                    count_kmers[kmer] += 1 # add the current key to the next key\n",
    "                else: count_kmers[kmer] = 1 # check if key is not in dictionary then count the current key only\n",
    "    print(count_kmers)\n",
    "count_kmers2unknown(fastq) # count_kmers2unknown function is run with fastq data\n",
    "\n",
    "\n",
    "# 4. Alter the counting function to allow for counting kmers of any size.\n",
    "def count_kmersanysize(file_name,n): # define a function with name 'count_kmersanysize', file_name refers to the data, n refers to kmersize\n",
    "    count_kmers = {}\n",
    "#check if key is in dictionary or not\n",
    "    for line_num,line in enumerate(file_name): # include the line number using \"enumerate\" so we can find our sequences\n",
    "        if line_num % 4 == 1: # if line number / 4 leaves a remainder of 1, we can get our sequence data and we can look for patterns   \n",
    "            for base_num,base in enumerate (line[0:-n]):\n",
    "                kmer = line[base_num:base_num+n]\n",
    "                kmer = ''.join(kmer)  # The method join() returns a string in which the string elements of sequence have been joined by str separator.\n",
    "                if kmer in count_kmers: # check if key is in dictionary\n",
    "                    count_kmers[kmer] += 1 # add the current key to the next key\n",
    "                else: count_kmers[kmer] = 1 # check if key is not in dictionary then count the current key only\n",
    "    print(count_kmers)\n",
    "count_kmersanysize(fastq,1) # count_kmersanysize function is run with fastq data for kmersize 1\n",
    "count_kmersanysize(fastq,2) # count_kmersanysize function is run with fastq data for kmersize 2\n",
    "count_kmersanysize(fastq,3) # count_kmersanysize function is run with fastq data for kmersize 3"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
